reparse.go

```diff
@@ -129,3 +140,69 @@ func EncodeReparsePoint(rp *ReparsePoint) []byte {
 	binary.Write(&b, binary.LittleEndian, target16)
 	return b.Bytes()
 }
+
+// ReparseDataBuffer corresponds to REPARSE_DATA_BUFFER struct defined here:
+// https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/ns-ntifs-_reparse_data_buffer.
+// ReparesPoint struct (and corresponding Encode/Decode functions) provided above only
+// handle symlink or mountpoint type of reparse points. To work with other reparse points
+// use this struct.  When calling SetReparsePoint, if it is a symlink or mountpoint
+// reparse point pass in the buffer generated by EncodeReparsePointData. If it is some
+// other kind of reparse point pass in the buffer generated by
+// ReparseDataBuffer.Encode. (same goes for GetReparsePoints)
+type ReparseDataBuffer struct {
+	ReparseTag        uint32
+	ReparseDataLength uint16
+	Reserved          uint16
+	// a generic data buffer
+	DataBuffer []byte
+}
+
+// Encode encodes the current ReparseDataBuffer struct that can be passed
+// to the SetReparsePoint.
+func (r *ReparseDataBuffer) Encode() []byte {
+	var buf bytes.Buffer
+	binary.Write(&buf, binary.LittleEndian, r.ReparseTag)
+	binary.Write(&buf, binary.LittleEndian, r.ReparseDataLength)
+	binary.Write(&buf, binary.LittleEndian, r.Reserved)
+	buf.Write(r.DataBuffer)
+	return buf.Bytes()
+}
+
+// Decode decodes the byte array (returned by GetReparsePoint) into
+// the current ReparseDataBuffer struct.
+func (r *ReparseDataBuffer) Decode(data []byte) error {
+	rdr := bytes.NewReader(data)
+	err := binary.Read(rdr, binary.LittleEndian, r)
+	if err != nil {
+		fmt.Errorf("failed to decode reparse data buffer: %w", err)
+	}
+	return nil
+}
+
+// GetReparsePoint returns a byte array that represents a ReparseDataBuffer if
+// the file at `path` is a valid reparse point.
+func GetReparsePoint(path string) ([]byte, error) {
+	utf16DestPath := windows.StringToUTF16(path)
```

Comment:
Prefer `windows.UTF16PtrFromString`

Reply:
I have changed this to `UTF16FromString`. I don't think we need to use `UTF16PtrFromString` but let me know if you think otherwise.